libutils/threaded_deque.c

```diff
@@ -0,0 +1,641 @@
+
+    if (size > 0)
+    {
+        data = xcalloc(size, sizeof(void *));
+        size_t right = deque->right;
+
+        for (size_t i = 0; i < size; i++)
+        {
+            right = right == 0 ? deque->capacity - 1 : right - 1;
+            data[i] = deque->data[right];
+            deque->data[right] = NULL;
+        }
+
+        deque->right = right;
+        deque->size -= size;
+    }
+
+    if (deque->size == 0)
+    {
+        // Signals that the deque is empty for ThreadedDequeWaitEmpty
+        pthread_cond_broadcast(deque->cond_empty);
+    }
+
+    *data_array = data;
+
+    ThreadUnlock(deque->lock);
+
+    return size;
+}
+
+size_t ThreadedDequePushLeft(ThreadedDeque *deque, void *item)
+{
+    assert(deque != NULL);
+
+    ThreadLock(deque->lock);
+
+    ExpandIfNecessary(deque);
+
+    deque->left = deque->left == 0 ? deque->capacity - 1 : deque->left - 1;
+    deque->data[deque->left] = item;
+    deque->size++;
+    size_t const size = deque->size;
+    pthread_cond_signal(deque->cond_non_empty);
+
+    ThreadUnlock(deque->lock);
+
+    return size;
+}
+
+size_t ThreadedDequePushRight(ThreadedDeque *deque, void *item)
+{
+    assert(deque != NULL);
+
+    ThreadLock(deque->lock);
+
+    ExpandIfNecessary(deque);
+
+    deque->data[deque->right++] = item;
+    deque->right %= deque->capacity;
+    deque->size++;
+    size_t const size = deque->size;
+
+    pthread_cond_signal(deque->cond_non_empty);
+
+    ThreadUnlock(deque->lock);
+
+    return size;
+}
+
+size_t ThreadedDequeCount(ThreadedDeque const *deque)
+{
+    assert(deque != NULL);
+
+    ThreadLock(deque->lock);
+    size_t const count = deque->size;
+    ThreadUnlock(deque->lock);
+
+    return count;
+}
+
+size_t ThreadedDequeCapacity(ThreadedDeque const *deque)
+{
+    assert(deque != NULL);
+
+    ThreadLock(deque->lock);
+    size_t const capacity = deque->capacity;
+    ThreadUnlock(deque->lock);
+
+    return capacity;
+}
+
+bool ThreadedDequeIsEmpty(ThreadedDeque const *deque)
+{
+    assert(deque != NULL);
+
+    ThreadLock(deque->lock);
+    bool const empty = (deque->size == 0);
+    ThreadUnlock(deque->lock);
+
+    return empty;
+}
+
+bool ThreadedDequeWaitEmpty(ThreadedDeque const *deque, int timeout)
+{
+    assert(deque != NULL);
+    int ret = true;
+
+    ThreadLock(deque->lock);
+
+    if (deque->size != 0)
+    {
+        if (timeout != 0)
+        {
+            int res = 0;
+
+            do {
+                res = ThreadWait(deque->cond_empty, deque->lock, timeout);
+
+                if (res != 0)
+                {
+                    /* Lock is reacquired even when timed out, so it needs to
+                       be released again. */
+                    ThreadUnlock(deque->lock);
+                    return false;
+                }
+            } while (deque->size != 0);
+            // Reevaluate predicate to protect against spurious wakeups
+        }
+        else
+        {
+            ret = false;
+        }
+    }
```

Comment:
I think this would be nicer if it was restructured to deal with the "easy" cases first. So things like
```
if (deque->size == 0)
{
    ThreadUnlock(deque->lock);
    return true;
}
```
and similarly for `timeout`.

Reply:
I assume you mean that the function should be "flatter" as a result, right?

Reply:
Exactly.